Add SubmissionModeration.update_crowd_control_level

MaybeNetwork opened the pull request:
## Feature Summary and Justification

Adds :meth:`.SumissionModeration.update_crowd_control_level`, which allows moderators to toggle the crowd control level of a single post. The table of level descriptions was taken from [here](https://mods.reddithelp.com/hc/en-us/articles/360038129231-Crowd-Control-).

This method raises a generic 403 when it is used on posts you don't have permissions for, or on posts that doesn't exist.

If you set `level` to some integers below 0, it sets the level to 0, and similarly, using some integers above 4 will set the level to 3. 

## References

* https://www.reddit.com/dev/api#POST_api_update_crowd_control_level
* https://mods.reddithelp.com/hc/en-us/articles/360038129231-Crowd-Control-

LilSpazJoekp commented on tests/integration/models/reddit/test_submission.py:
```diff
@@ -537,3 +537,19 @@ def test_unspoiler(self):
+                )
+            )
+            assert modlog.action == "adjust_post_crowd_control_level"
+            assert modlog.target_fullname == submission.fullname
```
```suggestion
            assert modlog.target_fullname == submission.fullname
            assert modlog.details == "medium"
```
Can we check that it actually set it to the correct level?

LilSpazJoekp reviewed (changes requested):
Looks great, just a couple of suggestions and questions.

1. How would this behave if a subreddit has global crowd control disabled? Would it still enforce it?
2. How would one revert the override for the post in question?
3. What happens if a level is sent that is outside of 0-3 like -1 or 4?

MaybeNetwork commented on praw/models/reddit/redditor.py:
```diff
@@ -203,11 +203,11 @@ def block(self):
 
-        For example, to remove Redditor ``spez`` from your whitelist:
+        For example, to remove Redditor ``jc7254`` from your whitelist:
 
         .. code-block:: python
 
-            reddit.redditor("spez").distrust()
+            reddit.redditor("jc7264").distrust()
```
Sort of.  When I made the initial PR I just picked names of early Reddit employees. In hindsight, it's not such a good idea to have the CEO as someone to "distrust". So I replaced his name with Jose Canseco's, a baseball player who did a famously bad  [AMA](https://old.reddit.com/r/IAmA/comments/1clw9o/i_am_jose_canseco_famed_steroid_user_and_former/) years ago. I think [this comment](https://old.reddit.com/r/IAmA/comments/1clw9o/i_am_jose_canseco_famed_steroid_user_and_former/c9hs7hr/) is a good example of why his account is a good fit.

LilSpazJoekp replied:
I see and I agree that it could be interpreted negatively. I think we should come up with an example username and use it globally. 

For now, let's spilt this to it's own PR. I'd actually like to use a username that is suspended for the distrust method specifically. There are spam/bad actor accounts out there that are suspended but preserved for public reference.

In the meantime, I'll poke @bboe and @jarhill0 for their thoughts on using a set example username, subreddit, submission, etc. to be used in docs because of the connotations￼ some methods could convey. I think we've briefly discussed it but nothing was decided.

MaybeNetwork replied:
> I see and I agree that it could be interpreted negatively. I think we should come up with an example username and use it globally.
> 
> For now, let's spilt this to it's own PR. I'd actually like to use a username that is suspended for the distrust method specifically. There are spam/bad actor accounts out there that are suspended but preserved for public reference.
> 
> In the meantime, I'll poke @bboe and @jarhill0 for their thoughts on using a set example username, subreddit, submission, etc. to be used in docs because of the connotations￼ some methods could convey. I think we've briefly discussed it but nothing was decided.

I can take it out. For what it's worth, the example for block also uses spez. Blocking employees or removing them from your whitelist doesn't actually do anything, since they can send you a message regardless of what your settings are.

LilSpazJoekp replied:
True, which is another reason we should find a good example username to use in examples. Having an employee as the target for the `block` example implies that one could successfully block an reddit employee which isn't true.

bboe replied:
This is an interesting discussion about usernames. I personally am impartial about the names in the documentation so long as anyone wouldn't be annoyed by people running said examples (e.g., I do not want my account to be messaged by people testing). IMO Using the co-creator of Reddit I think is reasonable, nevertheless, I support whatever @LilSpazJoekp recommends.

MaybeNetwork commented on praw/models/reddit/submission.py:
```diff
@@ -341,6 +341,39 @@ def unspoiler(self):
         """
         self.thing._reddit.post(API_PATH["unspoiler"], data={"id": self.thing.fullname})
 
+    def update_crowd_control_level(self, level: Union[str, int]):
```
I annotated it that way because it reflects that fact that it doesn't matter if you the data is  `{"level": 1}` or `{"level": "1"}`, because the Requests [formats](https://github.com/praw-dev/praw/pull/1753/files#diff-9b2db9631b5b5e5a8571dc9f43632b1d2196f79dc2a274b23e9bc926b51f5b98R109) the two in the same way. On the other hand, users that type check their scripts before running them might want it to just be an int, so I could change it if you'd like.

LilSpazJoekp replied:
Please, just because that is what reddit has documented for this endpoint.

MaybeNetwork commented:
I made the suggested changes to the tests, and switched the type annotation. I replaced the instances of "your community" in the descriptions with "the subreddit", and added my own description to the "off" setting. I had left the last item out because Reddit's description for the off setting is "Use Crowd Control to automatically collapse comments from users you're not sure about.", which doesn't describe what the setting does. I also added a link to the CommentModeration.show, which uncollapses comments that Crowd Control has actioned.

Is there an issue with how lists are displayed on ReadTheDocs? The bullets show up on the local build that pre_push.py generates, but the ReadTheDocs version doesn't show bullets. See https://archive.is/tdbyE

LilSpazJoekp commented on praw/models/reddit/submission.py:
```diff
@@ -341,6 +341,43 @@ def unspoiler(self):
+
+            submission = reddit.submission(id="745ryj")
+            submission.mod.update_crowd_control_level(2)
+
```
Could we add a note that if integers outside 0-3 are passed reddit will use the closest valid integer.

MaybeNetwork replied:
I'm hesitant to do this because it's undocumented behavior that doesn't really have a use. It's just something I came across during testing.

LilSpazJoekp replied:
Good point. I think it would be fine as long as we added a test for it and say in the note the behavior is what currently happens and could throw an exception in the future. I'm fine with it either way, though its current behavior is unexpected to me as I'd expect an error thrown when an invalid level is passed.

@bboe what do you think?

bboe replied:
I don't think PRAW should do anything specifically here. We don't check that usernames or subreddit names are allowable names and instead rely upon Reddit to perform that validation. I like having that level of simplicity in our library. Yes, that means something could break for someone who is using a value outside of the documented range, but if they're using a value outside of the documented range then the resulting behavior could reasonably be unexpected.

LilSpazJoekp replied:
I agree that we shouldn't do client side validation. My suggestion was for documenting Reddit's behavior to use the closest valid integer in this instance.

bboe replied:
I don't think we should mention that behavior lest it change.

LilSpazJoekp replied:
Sounds good!

LilSpazJoekp commented on tests/integration/models/reddit/test_submission.py:
```diff
@@ -537,3 +537,18 @@ def test_unspoiler(self):
+                )
+            )
+            assert modlog.action == "adjust_post_crowd_control_level"
+            assert modlog.details == "medium"
```
Nice catch on the alphabetical order here 🙂

LilSpazJoekp reviewed (changes requested):
One more thing I forgot to suggest. I missed the last part in the initial PR description.

LilSpazJoekp commented:
> Is there an issue with how lists are displayed on ReadTheDocs? The bullets show up on the local build that pre_push.py generates, but the ReadTheDocs version doesn't show bullets. See https://archive.is/tdbyE￼

Interesting. That is really strange. I'll investigate that further.

LilSpazJoekp reviewed (approved):
I'm going to go ahead and merge this. We can always add that note later if it causes any issues.

LilSpazJoekp commented:
Thank you! 🎇
